mv/plan/src/main/spark3.1/org/apache/carbondata/mv/plans/modular/SparkVersionHelper.scala

```diff
@@ -0,0 +1,166 @@
+    var attributeStats = AttributeMap(attributes.iterator
+      .map { pair => (rewrites(pair._1), pair._2) }.toSeq)
+    if (aliasMap.isDefined) {
+      attributeStats = AttributeMap(
+        attributeStats.map(pair => (aliasMap.get(pair._1), pair._2)).toSeq)
+    }
+    Statistics(stats.sizeInBytes, stats.rowCount, attributeStats)
+  }
+
+  def getOptimizedPlan(s: SubqueryExpression): LogicalPlan = {
+    val Subquery(newPlan, _) = BirdcageOptimizer.execute(Subquery.fromExpression(s))
+    newPlan
+  }
+
+  def normalizeExpressions[T <: Expression](r: T, attrs: AttributeSeq): T = {
+    QueryPlan.normalizeExpressions(r, attrs)
+  }
+
+  def attributeMap(rAliasMap: AttributeMap[Attribute]) : AttributeMap[Alias] = {
+    rAliasMap.asInstanceOf[AttributeMap[Alias]]
+  }
+
+  def seqOfRules : Seq[Rule[LogicalPlan]] = {
+    Seq(
+      // Operator push down
+      PushProjectionThroughUnion,
+      ReorderJoin,
+      EliminateOuterJoin,
+      PushPredicateThroughJoin,
+      PushDownPredicates,
+      ColumnPruning,
+      // Operator combine
+      CollapseRepartition,
+      CollapseProject,
+      CollapseWindow,
+      CombineFilters,
+      EliminateLimits,
+      CombineUnions,
+      // Constant folding and strength reduction
+      NullPropagation,
+      FoldablePropagation,
+      ConstantFolding,
+      ReorderAssociativeOperator,
+      // No need to apply LikeSimplification rule while creating MV
+      // as modular plan asCompactSql will be set in schema
+      //        LikeSimplification,
+      BooleanSimplification,
+      SimplifyConditionals,
+      RemoveDispensableExpressions,
+      SimplifyBinaryComparison,
+      EliminateSorts,
+      SimplifyCasts,
+      SimplifyCaseConversionExpressions,
+      RewriteCorrelatedScalarSubquery,
+      EliminateSerialization,
+      RemoveRedundantAliases)
+  }
+
+}
+
+trait getVerboseString extends LeafNode {
+  def verboseString: String = toString
+}
+
+trait groupByUnaryNode extends UnaryNode {
+  override def verboseString(maxFields: Int): String = super.verboseString(maxFields)
+
+  override def mapProductIterator[B](f: Any => B)(implicit evidence$1: ClassTag[B]): Array[B] = {
+    super.mapProductIterator(f)
+  }
+
+  override def mapChildren(f: ModularPlan => ModularPlan) : ModularPlan = {
+    val groupBy = super.mapChildren(f)
+    if (this.rewritten && !groupBy.rewritten) {
+      groupBy.setRewritten()
+    }
+    groupBy
+  }
+}
+
+trait selectModularPlan extends ModularPlan {
+  override def verboseString(maxFields: Int): String = super.verboseString(maxFields)
+
+  override def mapProductIterator[B](f: Any => B)(implicit evidence$1: ClassTag[B]): Array[B] = {
+    super.mapProductIterator(f)
+  }
+
+  override def mapChildren(f: ModularPlan => ModularPlan) : ModularPlan = {
+    val select = super.mapChildren(f)
+    if (this.rewritten && !select.rewritten) {
+      select.setRewritten()
+    }
+    select
+  }
+}
+
+trait unionModularPlan extends ModularPlan {
```

Comment:
please change all trait names, according to above @kunal642 comment

Reply:
done